### PR TITLE
extendWithInterface and connectToStore

### PR DESCRIPTION
**Pull request by Hugheth**

### CODE REVIEW

File: lib/ClassUtils.lua [hunk 1]
```diff
@@ -29,6 +29,32 @@ function ClassUtils.makeClass(name, constructor)
...
+					return input
+				end
+			end
+		end
+		local inheritedInterface = self.interface
+		local compositeInterface = function(Class)
+			return TableUtils.assign(
+				{},
+				getComposableInterface(interface)(Class),
+				getComposableInterface(inheritedInterface)(Class)
```
Comment (ckdavid) [inline lib/ClassUtils.lua]: What's the expected behaviour if `interface` and `inherited` interface have overlapping keys? If I extend a class A and make the interface for the subclass B less restrictive, I can now create instances of A that don't adhere to A's interface. Do we care?

File: lib/connectToStore.lua [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+		{
+			store = tea.interface(
+				{
+					changed = tea.Signal
+				}
+			)
+		}
+	)
+
+	function ConnectedClass:listen()
```
[the remark this answers is not in the document]
Comment (Hugheth) [reply lib/connectToStore.lua]: Could just be called mount instead?

The idea is to separate component lifecycle from bound lifecycle, for example if you wanted to create the component structure before it should be bound.

Haven't needed that yet though

File: lib/connectToStore.lua [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+		self._connection =
+			self._store.changed:connect(
+			function(state)
+				local nextFields = mapStateToFields(state)
```
[the remark this answers is not in the document]
Comment (Hugheth) [reply lib/connectToStore.lua]: the injected fields should be readonly like props - do you have a preferred way of representing that?

File: lib/ClassUtils.lua (hunk 1 above)
Comment (Hugheth) [reply lib/ClassUtils.lua]: Good spot. Not sure it matters too much, but I'll add a comment

File: lib/ClassUtils.lua [hunk 4]
```diff
@@ -29,6 +29,32 @@ function ClassUtils.makeClass(name, constructor)
...
+			else
+				return function()
+					return input
+				end
+			end
+		end
+		local inheritedInterface = self.interface
+		local compositeInterface = function(Class)
+			return TableUtils.assign(
+				{},
```
[the remark this answers is not in the document]
Comment (Hugheth) [reply lib/ClassUtils.lua]: it's not doing any harm, and saves mutating the input interface arg

File: lib/connectToStore.lua (hunk 3 above)
[the remark this answers is not in the document]
Comment (ckdavid) [reply lib/connectToStore.lua]: If they're meant to be like props, would it make sense to call them props?

[the remark this answers is not in the document]
Comment (Hugheth) [reply lib/connectToStore.lua]: not sure how that would prevent accidental mutation?

[the remark this answers is not in the document]
Comment (ckdavid) [reply lib/connectToStore.lua]: Just helping to make it clearer how they're expected to be used. If we wanted to do better, we'd probably need some sort of immutability library to wrap the data. Not sure if it's worth it though unless there's a clear good option. Maybe Owen would have ideas?

File: spec/ClassUtils_spec.lua [hunk 5]
```diff
@@ -638,6 +638,47 @@ bad value for key amount:
...
+							ClassUtils.makeFinal(
+							{
+								a = 2
+							}
+						)
+						assert.errors(
+							function()
+								myObject.b = 2
+							end,
+							"Attempt to set key b on final object"
```
Comment (ckdavid) [inline spec/ClassUtils_spec.lua]: What about a test of setting a field that is already defined (i.e. `a`)?

File: lib/ClassUtils.lua [hunk 6]
```diff
@@ -129,13 +129,23 @@ function ClassUtils.applySwitchStrategyForEnum(enum, enumValue, strategies, ...)
...
 
-function ClassUtils.makeSymbolEnum(keys)
-	return TableUtils.map(
-		ClassUtils.makeEnum(keys),
-		function(key)
-			return ClassUtils.makeSymbol(key)
+function ClassUtils.makeFinal(object)
+	local backend = getmetatable(object)
+	local proxy = {
+		__index = function(t, key)
```
Comment (ckdavid) [inline lib/ClassUtils.lua]: Will we ever iterate over these objects using `ipairs` or `pairs`. Apparently that won't work: http://lua-users.org/wiki/ReadOnlyTables. Worth testing?

**Pull request merged**